### ISSUE TITLE
Cube transpose behaviour with multidimensional coordinates

### ISSUE THREAD

**Issue by cpelley** [SciTools/iris#2606]
The context of this ticket is in utilising the capability from iris.experimental.stratify which necessarily requires us to transpose cubes to make them suitable for use.

The current implementation of [Cube.transpose](https://github.com/SciTools/iris/blob/6bd26b779bf5da202e933a6bf4d87c8c7eb0160a/lib/iris/cube.py#L2809) changes the dimension mapping of the axillary coordinates, but doesn't actually transpose the coordinates themselves.  This is so confusing! (see the following [gist](https://gist.github.com/cpelley/5dcd0715efed2869ffeb42ef46385d68)).

Note that the altitude coordinate likely behaves in the way that I was expecting, unlike the surface altitude as it is a derived coordinate... which only adds to the confusion with the differing shapes...

I would appreciate a second opinion on this, cheers.
I would consider this a bug on the basis that I was not expecting this at all and I'm interested in hearing what others think.

**Comment by cpelley** [comment]
Additionally, the above would then effect aux coord equality.

**Comment by ajdawson** [comment]
I've encountered this before and it caught me by surprise. I'd call it a bug.

**Comment by cpelley** [comment]
Thanks for your input @ajdawson.  I'm glad I'm not the only one :)

**Comment by cpelley** [comment]
Looking at the recent [iris.experimental.stratify](https://github.com/SciTools/iris/blob/6bd26b779bf5da202e933a6bf4d87c8c7eb0160a/lib/iris/experimental/stratify.py) implementation, it looks as though it assumed that dimension mapping for aux coords is ordered increasing.  I wonder how much other iris code also assumes this to be the case...

**Comment by rcomer** [comment]
Possibly worth noting that `iris.util.as_compatible_shape` treats any transposed auxcoords the same way as `cube.transpose` (#2030).  So if it's a bug, then it's a bug in at least two places!

**Comment by cpelley** [comment]
Thanks @rcomer, I didn't realise.  Agreed, we should ensure the two remain consistent in their whatever we do :)

Perhaps the approach to resolving this is to list the options available to us and then can consider ranking them in order of least to greatest confusion/usefulness.

Here are the options as I see them:

1. Leave the current behaviour where aux coords associated dims are re-arranged when the cube is transposed.
2. Transpose aux coords like we do with the cube.
3. Transpose aux coords when transposing cube but ensure that this is done in such a way that the ordered mappings are ordered increasing (i.e. [0, 3, 2] becomes [0, 2, 3]).

I hope I have explained these options clearly enough :)

**Comment by cpelley** [comment]
After spending quite some time on this, I think on transposing the cube at least, it might be most useful to take the approach of pnt3.  The reason being, is that the mappings of the coordinate might not originally be ordered increasing and so simply transposing this won't make things any clearer.  By enforcing increasing order of the coordinate dimension mappings when requesting a transpose, we have a consistent and obvious behaviour.  However I'm agnostic as to whether this makes sense in the context of `as_compatible_shape` as here it relates between another cube.

**Comment by cpelley** [comment]
I have put up a [PR against my own fork](https://github.com/cpelley/iris/pull/38) to demonstrate what the changes might look like, so we can explore this.

**Comment by rcomer** [comment]
I think that intuitively you'd expect auxcoord dims to be increasing, so option 3 sounds good to me.  It's never been a problem as such, but was quite confusing the first time I encountered it with `cube.transpose`.

As for `as_compatible_shape`, I've never used it directly (when it's called by `iris.analysis.maths._binary_op_common`, only the cube data is kept anyway).

**Comment by cpelley** [comment]
Thanks @rcomer.  Something that just occurred to me.  Perhaps this is a behavioural change worth considering across the board i.e. when attaching a coordinate to a cube as well as these situation where we are reordering a cube.

**Comment by rcomer** [comment]
I'd worry that it might be difficult to find and pin down every possible way for the auxcoords to end up with "wrong" dimension order.  Maybe they are that way when you first load the cube from a file.

An alternative might be to write some sort of `sanitize_auxcoords` function, which could be applied to make sure they're in the expected order when it matters (e.g. the `stratify` function).

But we're getting into code design now, which is not my area of expertise...!

**Comment by cpelley** [comment]
> Maybe they are that way when you first load the cube from a file.

Loading a cube from disk would mean the format in question going through cube construction though :)

+1 on a helper function though (this is what I have locally now).  Great name for it too!

**Comment by cpelley** [comment]
I have just added a `sanitise_auxcoords` function to illustrate in [my branch here](https://github.com/cpelley/iris/pull/38/files).  cube.transpose can call the sanitise function if we do go that route.

**Comment by pelson** [comment]
To be clear, the discussion here is about whether a cube should allow coordinates to be added where the dimensions aren't ordered, right?

NetCDF has no such constraint, but I'm uncertain about the CF conventions. Personally, I'm OK with having the order of the dimensions be non monotonic, but can see that there are some situations where that does make some code harder to implement. I'm partially tempted to suggest a normalisation function which allows non monotonically increasing coordinate dims to be made monotonic (and thus transposing the coordinates as necessary).

@cpelley - at this point, I'm looking for a good reason to justify the monotonic dimensions constraint, otherwise I would not be in favour of the change. Are you able to help?

**Comment by cpelley** [comment]
> @cpelley - at this point, I'm looking for a good reason to justify the monotonic dimensions constraint, ...

Not quite.  It was proposed that a cube could automatically ensure that coordinates have monotonic dimension ordering as driven by the confusion that arises otherwise - see above (transpose in particular).
Alternatively, just transpose and as_compatible_shape could be made to transpose the coordinates themselves and not the current behaviour of changing the dimension mapping).
If pnt2 or neither point above takes your fancy, then adding a `sanitise_auxcoords` utility to iris would at least allow people to address these concerns in a manual way (I have done this already for ANTS).

Cheers

**Comment by pelson** [comment]
> If pnt2 or neither point above takes your fancy, then adding a sanitise_auxcoords utility to iris would at least allow people to address these concerns in a manual way

I'd be happy to review a PR that adds a ``ensure_auxcoord_dims_monotonic`` function (maybe in ``iris.util``).
